### PR TITLE
[CPU] Add tests for SetBlob + I64 support

### PR DESCRIPTION
**Pull request by mandrono**
\+ Issue: 34516

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
Common question: can such tests be applied for other plugin as well? Looks like we need to move them to shared_tests. 
BTW, do shared_tests already have some tests for setBlob? I suppose yes

**Comment by mandrono** [comment]
> Common question: can such tests be applied for other plugin as well? Looks like we need to move them to shared_tests.
> BTW, do shared_tests already have some tests for setBlob? I suppose yes

As far as I know, we have no tests that check correct work for setBlob with layout and precision different from network layout and precision.

**Comment by ilya-lavrenov** [comment]
> > Common question: can such tests be applied for other plugin as well? Looks like we need to move them to shared_tests.
> > BTW, do shared_tests already have some tests for setBlob? I suppose yes
> 
> As far as I know, we have no tests that check correct work for setBlob with layout and precision different from network layout and precision.

Ok, that means we can add these tests for all plugins to a shared folder, right?

**Comment by mandrono** [comment]
@dmitry-gorokhov @AlexPeskov review please

**Comment by ilya-lavrenov** [comment]
can we create tasks for other plugins to enable such tests on their side?

File: inference-engine/tests/functional/plugin/shared/include/behavior/set_blob.hpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+using setBlobParams = std::tuple<InferenceEngine::Precision,   // input precision
+                                 InferenceEngine::Layout,      // input layout
+                                 InferenceEngine::Precision,   // output precision
+                                 InferenceEngine::Layout,      // output layout
+                                 setType,                      // type for which blob is set
+                                 InferenceEngine::SizeVector,  // input shape
+                                 std::string>;                 // Device name
+
+class setBlobTest : public testing::WithParamInterface<setBlobParams>, virtual public LayerTestsUtils::LayerTestsCommon {
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/include/behavior/set_blob.hpp:27]: Why do you use `LayerTestsUtils::LayerTestsCommon`? It could be unjustifiably.
Furthermore use `SetBlobTest ` and `SetBlobParams ` instead of `setBlobTest ` and `setBlobParams `. These names are type names

File: inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.cpp [hunk 2]
```diff
@@ -140,13 +152,19 @@ std::vector<std::vector<std::uint8_t>> LayerTestsCommon::CalculateRefs() {
...
     if (outPrc == InferenceEngine::Precision::UNSPECIFIED) {
-        const auto &actualOutputs = GetOutputs();
-        ieOutPrc = actualOutputs[0]->getTensorDesc().getPrecision();
+        for (size_t i = 0; i < ieOutPrc.size(); i++) {
+            ieOutPrc[i] = actualOutputs[i]->getTensorDesc().getPrecision();
+        }
+    }
+
+    std::vector<ngraph::element::Type_t> convertType(ieOutPrc.size());
+    for (size_t i = 0; i < ieOutPrc.size(); i++) {
```
Comment (iefode) [inline inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.cpp:164]: you can merge both cycles. Looks like duplication. And What is the reason of creation these vectors? I don't see using these variables.

File: inference-engine/tests/ngraph_functions/src/utils/ngraph_helpers.cpp [hunk 3]
```diff
@@ -169,8 +171,8 @@ std::vector<std::vector<std::uint8_t>> getConstData(const std::shared_ptr<Functi
         const auto dataSize = shape_size(parrentNode->get_shape()) * parrentNode->get_element_type().size();
         outputs[i].resize(dataSize);
         std::copy(data, data + dataSize, outputs[i].data());
-        if (convertType != element::Type_t::undefined)
-            outputs[i] = convertOutputPrecision(outputs[i], parrentNode->get_element_type(), convertType, shape_size(parrentNode->get_shape()));
+        if (!convertType.empty() && convertType[i] != element::Type_t::undefined)
```
Comment (iefode) [inline inference-engine/tests/ngraph_functions/src/utils/ngraph_helpers.cpp:174]: if `convertType.empty()` is `true`, we won't enter cycle.

**Review by iefode — CHANGES_REQUESTED** [review]

File: inference-engine/tests/functional/plugin/shared/include/behavior/set_blob.hpp (hunk 1 above)
Comment (mandrono) [reply inference-engine/tests/functional/plugin/shared/include/behavior/set_blob.hpp]: I believe we should compare result IE inference and NGrpah inference in this test

File: inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.cpp (hunk 2 above)
Comment (mandrono) [reply inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.cpp]: we need to create vectors of output types from IE network to convert NGraph output to their type
previous implementation convert all NGraph output to actualOutputs[0] type

File: inference-engine/tests/ngraph_functions/src/utils/ngraph_helpers.cpp (hunk 3 above)
Comment (mandrono) [reply inference-engine/tests/ngraph_functions/src/utils/ngraph_helpers.cpp:174]: ```convertType``` is empty in case if we don`t specify output precision and will be compare results as is

**Comment by mandrono** [comment]
@nshchego review please

File: inference-engine/tests/functional/plugin/shared/src/behavior/set_blob.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
+// Copyright (C) 2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "behavior/set_blob.hpp"
+#include <single_layer_tests/cum_sum.hpp>
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/src/behavior/set_blob.cpp:6]: Why? `#include <single_layer_tests/cum_sum.hpp>`

Comment (mandrono) [reply inference-engine/tests/functional/plugin/shared/src/behavior/set_blob.cpp:6]: I used CumSum layer to build ngraph fucntion because this layer supports 1D - 5D input tensor

Comment (mandrono) [reply inference-engine/tests/functional/plugin/shared/src/behavior/set_blob.cpp:6]: I just used CumSum layer to build ngraph fucntion

**Review by nshchego — APPROVED** [review]

File: inference-engine/tests/functional/plugin/shared/src/behavior/set_blob.cpp (hunk 4 above)
Comment (iefode) [reply inference-engine/tests/functional/plugin/shared/src/behavior/set_blob.cpp:6]: As I can see, you use all defined methods from `CumSum` and overload some methods from 'LayerTestsCommon'. Maybe you should inherit directly from 'CumSumLayerTests'

**Review by iefode — APPROVED** [review]

**Comment by mandrono** [comment]
@AlexPeskov review please

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp [hunk 5]
```diff
@@ -234,10 +252,23 @@ void MKLDNNPlugin::MKLDNNInferRequest::GetBlob(const char *name, InferenceEngine
             return;
         }
 
-        _outputs[name] = make_blob_with_precision(blobs[name]->getTensorDesc());
+        InferenceEngine::TensorDesc desc = blobs[name]->getTensorDesc();
+
+        auto outNodes = graph->GetOutputNodes();
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp:257]: Code block below looks like WA. Please put a comment why we need it.

File: inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.cpp [hunk 6]
```diff
@@ -35,7 +35,11 @@ void LayerTestsCommon::Compare(const std::vector<std::uint8_t> &expected, const
     const auto actualBuffer = lockedMemory.as<const std::uint8_t *>();
 
     const auto &precision = actual->getTensorDesc().getPrecision();
-    const auto &size = actual->size();
+
+    size_t size = actual->size();
```
Comment (dmitry-gorokhov) [inline inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.cpp:39]: Why do you need these changes?

File: ngraph/python/tests/__init__.py [hunk 7]
```diff
@@ -45,6 +45,8 @@ def xfail_test(reason="Mark the test as expected to fail", strict=True):
                                       "DynamicQuantizeLinear")
 xfail_issue_33538 = xfail_test(reason="RuntimeError: nGraph does not support the following ONNX operations:"
                                       "Scan")
+skip_issue_38084 = pytest.mark.skip(reason="Aborted (core dumped) Assertion "
```
Comment (dmitry-gorokhov) [inline ngraph/python/tests/__init__.py:48]: Why you changed ngraph python tests?

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp (hunk 5 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp]: done

**Comment by mandrono** [comment]
@dmitry-gorokhov review please

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 8]
```diff
@@ -758,24 +768,57 @@ void MKLDNNGraph::PullOutputData(BlobMap &out) {
...
-        cpu_memcpy_s(ext_blob_ptr, ext_blob->byteSize(), intr_blob_ptr, size_to_copy);
+        auto srcPrec = MKLDNNMemory::convertToIePrec(intr_blob.GetDataType());
+        auto dstPrec = ext_blob->getTensorDesc().getPrecision();
+        if (dstPrec == srcPrec || (dstPrec == Precision::BOOL && srcPrec == Precision::U8)) {
+            if (ext_blob->byteSize() != intr_blob.GetSize())
+                THROW_IE_EXCEPTION << "Output blob byte size is not equal network output byte size ("
+                                   << ext_blob->byteSize() << "!=" << intr_blob.GetSize() << ").";
+
+            // That is the same memory. No need to copy
+            if (ext_blob_ptr == intr_blob_ptr) continue;
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:789]: Why this condition was moved inside `dstPrec == srcPrec` case? Should it be valid for any case?

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 9]
```diff
@@ -758,24 +768,57 @@ void MKLDNNGraph::PullOutputData(BlobMap &out) {
...
         int MB_to_process = node->batchToProcess();
         // TODO: Should we support InferenceEngine::PluginConfigParams::KEY_DYN_BATCH_LIMIT???
         if (config.batchLimit)
             MB_to_process = std::min<int>(config.batchLimit, MB_to_process);
         size_t size_to_copy = intr_blob.GetSize() * MB_to_process / MB;
 
-        cpu_memcpy_s(ext_blob_ptr, ext_blob->byteSize(), intr_blob_ptr, size_to_copy);
+        auto srcPrec = MKLDNNMemory::convertToIePrec(intr_blob.GetDataType());
+        auto dstPrec = ext_blob->getTensorDesc().getPrecision();
+        if (dstPrec == srcPrec || (dstPrec == Precision::BOOL && srcPrec == Precision::U8)) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:783]: What do you think about handling of `dstPrec == srcPrec` case inside convert itself? So if typename src == typename dst you will be able to call cpu_memcpy_s. This will allow you not to check such custom cases as `dstPrec == Precision::BOOL && srcPrec == Precision::U8`

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 10]
```diff
@@ -758,24 +768,57 @@ void MKLDNNGraph::PullOutputData(BlobMap &out) {
...
+
+            // That is the same memory. No need to copy
+            if (ext_blob_ptr == intr_blob_ptr) continue;
+
+            cpu_memcpy_s(ext_blob_ptr, ext_blob->byteSize(), intr_blob_ptr, size_to_copy);
+        } else {
+            if (srcPrec != Precision::FP32 && srcPrec != Precision::I32)
+                THROW_IE_EXCEPTION << "Can't convert: unsupport network output precision: " << srcPrec;
+            const size_t sizeToConvert = size_to_copy / srcPrec.size();
+            switch (dstPrec) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:796]: Looks like it make sense to use `getPrecisionMask` routine here,

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp [hunk 11]
```diff
@@ -164,6 +162,24 @@ void MKLDNNPlugin::MKLDNNInferRequest::InferImpl() {
...
+                    auto in = dynamic_cast<InferenceEngine::TBlob<std::int32_t> *>(iconv.get());
+                    if (in == nullptr)
+                        THROW_IE_EXCEPTION << "Cannot get TBlob";
+                    if (input.second->getTensorDesc().getPrecision() == InferenceEngine::Precision::I64)
+                        copyFrom<int64_t, int32_t>(input.second.get(), in->data());
+                    else
+                        copyFrom<uint64_t, int32_t>(input.second.get(), in->data());
+                    pushInput<int32_t>(input.first, iconv);
+                    }
+                    break;
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp:182]: Could you please extract the code block above (lines 165 - 181) into separate method and reuse between other switch cases (U16, I16, BOOL).

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 12]
```diff
@@ -734,6 +734,16 @@ void MKLDNNGraph::PushInputData(const std::string& name, const InferenceEngine::
     }
 }
 
+template <typename src, typename dst>
+void convert(void *_dataSrc, void *_dataDst, const size_t &size) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:738]: Lets make it static

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 9 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:783]: If I understood your idea correctly, then it will look like this: we create list precision which we directly support in plugin (including ```BOOL```), if ```dstPrec``` is included in this list we call instance of ```convert```, else process as now.
I believe we don't need this. It is increase binary size and we don`t have many such specific cases like BOOL and U8

Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:783]: We can consider the option to extract ```Convert Layer``` execute function and use it if we need this, it's not increase binary size and we got more simple copy function

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 8 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp]: yes, it is. at this moment we change pointer for node on pointer to blob in case:
input if network input precision equals mkldnn graph input node precision
for other if precision equals fp32

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 10 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp (hunk 11 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp:152]: fixed

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 12 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+#include "cpu_memcpy.h"
+#include <type_traits>
+#include <ie_parallel.hpp>
+
+using namespace InferenceEngine;
+
+template<typename srcType, typename dstType>
+void convert(void *srcPtr, void *dstPtr, const size_t &size) {
+    if (std::is_same<srcType, dstType>::value)
+        cpu_memcpy(dstPtr, srcPtr, size*sizeof(dstType));
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp:15]: I think rest part of the function should be inside `else` branch otherwise you perform memory copy twice.

File: inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.hpp [hunk 14]
```diff
@@ -82,6 +82,26 @@ class LayerTestsCommon : public CommonTestUtils::TestsCommon {
         }
     }
 
+    // need to avoid error (taking the absolute value of unsigned type has no effect) on mac
+    template<class T>
```
Comment (dmitry-gorokhov) [inline inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.hpp:86]: Conditional template specialization will look better I suppose. See https://github.com/openvinotoolkit/openvino/blob/5f79c823f5bb7e2526f9ccfe1c31964761aac352/inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.hpp

File: inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+void cpu_convert(void *srcPtr, void *dstPtr, Precision srcPrc, Precision dstPrc, const size_t &size) {
+    if (srcPrc == dstPrc) {
+        cpu_memcpy(dstPtr, srcPtr, size*dstPrc.size());
+        return;
+    }
+
+    auto precMask = getPrecisionMask(srcPrc, dstPrc);
+    switch (precMask) {
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/common/cpu_memcpy.h [hunk 16]
```diff
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <stdlib.h>
+#include <cstring>
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/common/cpu_memcpy.h:7]: ```memcpy``` contains in this header without this include get error: ‘memcpy’ is not a member of ‘std’
+ remove #include <stdlib.h> I cant see any function from this header

File: inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.h [hunk 17]
```diff
@@ -0,0 +1,9 @@
+// Copyright (C) 2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <ie_precision.hpp>
+
+// @param size
+// number of elements to convert
+void cpu_convert(void *srcPtr, void *dstPtr, InferenceEngine::Precision srcPrc, InferenceEngine::Precision dstPrc, const size_t &size);
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.h]: done

File: inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp (hunk 13 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp:15]: thanks, it's my omission

File: inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.hpp (hunk 14 above)
Comment (mandrono) [reply inference-engine/tests/ie_test_utils/functional_test_utils/layer_test_utils.hpp:78]: fixed

File: inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp [hunk 18]
```diff
@@ -55,33 +56,24 @@ void MKLDNNPlugin::MKLDNNInferRequest::pushInput(const std::string& inputName, I
...
-        THROW_IE_EXCEPTION << "Input data was not allocated.";
-    }
-    for (size_t i = 0; i < t_blob->size(); i++) dst[i] = srcPtr[i];
-}
-
-template <typename T>
-void copyToFloat(float* dst, const InferenceEngine::Blob* src) {
-    copyFrom<T>(src, dst);
+template <typename dst>
+void MKLDNNPlugin::MKLDNNInferRequest::copyConvert(InferenceEngine::Precision convertTo, const std::pair<std::string, InferenceEngine::Blob::Ptr> &input,
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp:60]: Lets do in separate PR: please unite pushInput and copyConvert function. Dispatching on particular behavior should be inside.

**Review by dmitry-gorokhov — APPROVED** [review]
